src/nodes/mod.rs

```diff
@@ -20,9 +20,109 @@ pub use leaf::{LeafNode, LeafNodeRef};
+    Extension(ExtensionNode),
+    /// Variant representing a [LeafNode].
+    Leaf(LeafNode),
+}
+
+impl Encodable for TrieNode {
+    fn encode(&self, out: &mut dyn alloy_rlp::BufMut) {
+        match self {
+            Self::Branch(branch) => branch.encode(out),
+            Self::Extension(extension) => extension.encode(out),
+            Self::Leaf(leaf) => leaf.encode(out),
+        }
+    }
+
+    fn length(&self) -> usize {
+        match self {
+            Self::Branch(branch) => branch.length(),
+            Self::Extension(extension) => extension.length(),
+            Self::Leaf(leaf) => leaf.length(),
+        }
+    }
+}
+
+impl Decodable for TrieNode {
+    fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
+        let mut bytes = Header::decode_bytes(buf, true)?;
+
+        let mut items = vec![];
+        while !bytes.is_empty() {
+            // Decode header without advancing.
+            let Header { payload_length, .. } = Header::decode(&mut &bytes[..])?;
+            let len = if payload_length == 1 {
+                1 // If payload length is 1 byte, than there is no header
+            } else {
+                payload_length + length_of_length(payload_length)
+            };
+            items.push(bytes[..len].to_vec());
+            bytes.advance(len);
+        }
+
+        if items.len() == 17 {
+            let mut branch = BranchNode::default();
+            for (idx, item) in items.into_iter().enumerate() {
+                if idx == 16 {
+                    if item != [EMPTY_STRING_CODE] {
+                        return Err(alloy_rlp::Error::Custom(
+                            "branch node values are not supported",
+                        ));
+                    }
+                } else if item != [EMPTY_STRING_CODE] {
+                    branch.stack.push(item);
+                    branch.state_mask.set_bit(idx as u8);
+                }
+            }
+            return Ok(Self::Branch(branch));
+        }
+
+        if items.len() == 2 {
+            let key = items.remove(0);
+
+            let encoded_key = Bytes::decode(&mut &key[..])?;
+            if encoded_key.is_empty() {
+                return Err(alloy_rlp::Error::Custom("trie node key empty"));
+            }
+
+            let key_flag = encoded_key[0] & 0xf0;
+            // Retrieve first byte. If it's [Some], then the nibbles are odd.
+            let first = match key_flag {
+                0x10 | 0x30 => Some(encoded_key[0] & 0x0f),
+                0x00 | 0x20 => None,
+                _ => return Err(alloy_rlp::Error::Custom("node is not extension or leaf")),
+            };
+
+            let key = unpack_path_to_nibbles(first, &encoded_key[1..]);
+            let node = if key_flag == 0x20 || key_flag == 0x30 {
+                Self::Leaf(LeafNode::new(key, Bytes::decode(&mut &items.remove(0)[..])?.to_vec()))
+            } else {
+                Self::Extension(ExtensionNode::new(key, items.remove(0)))
```

Comment:
```suggestion
                // We don't decode this because we pull the value from the stack and it's already RLP Encoded
                Self::Extension(ExtensionNode::new(key, items.remove(0)))
```